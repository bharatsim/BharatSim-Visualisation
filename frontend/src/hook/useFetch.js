import { useEffect, useState } from 'react';

import useLoader from './useLoader';
import useDeepCompareMemoize from './useDeepCompareMemoize';

const defaultApiParameters = [];

function useFetch(api, apiParameters = defaultApiParameters, shouldFetch = true) {
  const [responseData, setResponseData] = useState();

  const { loadingState, startLoader, stopLoaderAfterError, stopLoaderAfterSuccess } = useLoader();

  const memoizeValue = useDeepCompareMemoize(apiParameters);

  useEffect(() => {
    if (!shouldFetch) {
      stopLoaderAfterSuccess();
      return;
    }
    startLoader();
    async function fetchApiData() {
      try {
        const resData = await api(...apiParameters);
        setResponseData(resData);
        stopLoaderAfterSuccess();
      } catch (e) {
        stopLoaderAfterError();
      }
    }

    fetchApiData();
  }, memoizeValue);

  return { data: responseData, loadingState: loadingState.state };
}

export default useFetch;
