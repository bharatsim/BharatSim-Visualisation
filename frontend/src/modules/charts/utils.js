import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
