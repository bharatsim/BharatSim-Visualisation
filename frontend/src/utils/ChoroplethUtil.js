import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
