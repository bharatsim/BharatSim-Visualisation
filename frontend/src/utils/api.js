import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
