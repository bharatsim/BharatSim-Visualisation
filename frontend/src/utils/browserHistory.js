// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
