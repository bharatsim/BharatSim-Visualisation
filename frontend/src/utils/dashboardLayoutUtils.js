export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
