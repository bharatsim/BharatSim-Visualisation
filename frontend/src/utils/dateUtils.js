const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function getPeriod(hours) {
  return hours >= 12 ? 'PM' : 'AM';
}

function formatDate(dateString) {
  const date = new Date(dateString);
  if (date.toString() === 'Invalid Date') return '--';
  const DD = date.getDate();
  const MM = date.getMonth();
  const YYYY = date.getFullYear();
  const HH = date.getHours();
  const mm = date.getMinutes();

  return `${DD}-${months[MM]}-${YYYY} at ${HH % 12}:${mm < 10 ? `0${mm}` : mm} ${getPeriod(HH)}`;
}

function currentDate(){
  return new Date();
}

export { formatDate, currentDate };
