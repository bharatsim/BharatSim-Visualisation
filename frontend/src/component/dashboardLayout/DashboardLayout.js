import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import { Box, Button, withStyles } from '@material-ui/core';
import ReactGridLayout, { WidthProvider } from 'react-grid-layout';

import 'react-grid-layout/css/styles.css';
import 'react-resizable/css/styles.css';
import styles from './dashboardLayoutCss';

import { getNewWidgetLayout, renderElement } from '../../utils/dashboardLayoutUtils';
import labels from '../../constants/labels';
import ChartConfigModal from '../chartConfigModal/ChartConfigModal';
import FileUpload from '../fileUpload/FileUpload';

import useFetch from '../../hook/useFetch';

import useModal from '../../hook/useModal';
import { api } from '../../utils/api';
import useLoader from '../../hook/useLoader';
import ChartSelector from '../chartSelector/ChartSelector';
import InlineLoader from '../loaderOrError/InlineLoader';
import LoaderOrError from '../loaderOrError/LoaderOrError';

const GridLayout = WidthProvider(ReactGridLayout);

const cols = 12;

function DashboardLayout({ classes }) {
  const [dashboardConfig, setDashboardConfig] = useState({
    name: 'dashboard1',
    id: null,
    count: 0,
  });
  const [widgets, setWidgets] = useState([]);
  const [layout, setLayout] = useState([]);
  const [chartType, setChartType] = useState();

  const { isOpen, closeModal, openModal } = useModal();
  const { data: allDashboards, loadingState: dashboardLoadingState } = useFetch(
    api.getAllDashBoard,
  );
  const { loadingState, startLoader, stopLoaderAfterError, stopLoaderAfterSuccess } = useLoader();

  useEffect(() => {
    if (allDashboards && allDashboards.dashboards.length > 0) {
      const firstDashboard = allDashboards.dashboards[0];
      const { name, _id, count } = firstDashboard;
      setWidgets(firstDashboard.widgets);
      setLayout(firstDashboard.layout);
      setDashboardConfig({ name, id: _id, count });
    }
  }, [allDashboards]);

  function addItem(config) {
    setWidgets((prevWidgets) => {
      return prevWidgets.concat({
        config,
        chartType,
        layout: getNewWidgetLayout(prevWidgets.length, cols, dashboardConfig.count),
      });
    });
    setDashboardConfig((prevState) => ({ ...prevState, count: prevState.count + 1 }));
  }

  function handleModalOk(config) {
    addItem(config);
    closeModal();
  }

  function onLayoutChange(changedLayout) {
    setLayout(changedLayout);
  }

  function onChartClick(selectedChartType) {
    openModal();
    setChartType(selectedChartType);
  }
  function onClickOfSaveDashboard() {
    startLoader(`Saving ${dashboardConfig.name}...`);
    api
      .saveDashboard({
        widgets,
        layout,
        dashboardId: dashboardConfig.id,
        name: dashboardConfig.name,
        count: dashboardConfig.count,
      })
      .then((data) => {
        setDashboardConfig((prevState) => ({
          ...prevState,
          id: data.dashboardId,
        }));
        stopLoaderAfterSuccess(`Dashboard ${dashboardConfig.name} Saved Successfully`);
      })
      .catch(() => {
        stopLoaderAfterError(`Failed to save dashboard ${dashboardConfig.name}`);
      });
  }

  return (
    <Box pl={10} pt={10} pr={10}>
      <Box pb={2}>
        <FileUpload />
      </Box>
      <Box pb={2} display="flex" justifyContent="space-between">
        <ChartSelector onClick={onChartClick} />
        <Box>
          <Button onClick={onClickOfSaveDashboard} variant="contained" color="primary">
            {labels.dashboardLayout.SAVE_DASHBOARD_BUTTON}
          </Button>
          <InlineLoader status={loadingState.state} message={loadingState.message} />
        </Box>
      </Box>
      <GridLayout
        layout={layout}
        onLayoutChange={onLayoutChange}
        className={classes.reactGridLayout}
      >
        {widgets.map((item) => {
          return renderElement(item);
        })}
      </GridLayout>
      {isOpen && (
        <ChartConfigModal
          onCancel={closeModal}
          onOk={handleModalOk}
          open={isOpen}
          chartType={chartType}
        />
      )}
    </Box>
  );
}

DashboardLayout.propTypes = {
  classes: PropTypes.shape({
    reactGridLayout: PropTypes.string.isRequired,
  }).isRequired,
};

export default withStyles(styles)(DashboardLayout);
