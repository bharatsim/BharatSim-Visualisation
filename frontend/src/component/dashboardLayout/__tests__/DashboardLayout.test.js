import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import DashboardLayout from '../DashboardLayout';

import useFetch from '../../../hook/useFetch';
import { api } from '../../../utils/api';

jest.mock('../../../hook/useFetch');
jest.mock('../../../utils/api', () => ({
  api: {
    saveDashboard: jest.fn(),
  },
}));

jest.mock('../../chartConfigModal/ChartConfigModal', () => ({ open, onCancel, onOk }) => (
  <>
    <span>
      Modal Open
      {`${open.toString()}`}
    </span>
    <button type="button" onClick={() => onOk('config')}>
      Ok
    </button>
    <button type="button" onClick={onCancel}>
      Cancel
    </button>
  </>
));

jest.mock('../../fileUpload/FileUpload', () => (props) => (
  <div>
    File Upload
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

jest.mock('../../../modules/charts/barChart/BarChart', () => (props) => (
  <div>
    Bar Chart
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

jest.mock('../../../modules/charts/lineChart/LineChart', () => (props) => (
  <div>
    Line Chart
    {/* eslint-disable-next-line no-undef */}
    <span>{mockPropsCapture(props)}</span>
  </div>
));

describe('<DashboardLayout />', () => {
  beforeEach(() => {
    useFetch.mockReturnValue({ data: { dashboards: [] }, loadingState: 'SUCCESS' });
    api.saveDashboard.mockResolvedValue({ dashboardId: 'id' });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match a snapshot for <DashboardLayout /> at initial state', async () => {
    const { container } = render(<DashboardLayout />);

    await act(async () => {
      expect(container).toMatchSnapshot();
    });
  });

  it('should match a snapshot for <DashboardLayout /> when dashboard is loaded with data', async () => {
    const data = {
      name: 'dashboard1',
      widgets: [
        {
          layout: { i: 'widget-0', x: 0, y: null, w: 2, h: 2 },
          config: 'config',
          chartType: 'barChart',
        },
      ],
      dashboardId: null,
      layout: [{ w: 2, h: 2, x: 0, y: 0, i: 'widget-0', moved: false, static: false }],
    };

    useFetch.mockReturnValue({ data: { dashboards: [data] }, loadingState: 'SUCCESS' });

    const { container } = render(<DashboardLayout />);

    expect(container).toMatchSnapshot();
  });

  it('should show loader while fetching data for <DashboardLayout />', async () => {
    useFetch.mockReturnValue({ data: undefined, loadingState: 'LOADING' });

    const { container } = render(<DashboardLayout />);

    expect(container).toMatchSnapshot();
  });

  it('should show Error while fetching data for <DashboardLayout /> if error occurred', async () => {
    useFetch.mockReturnValue({ data: undefined, loadingState: 'ERROR' });

    const { container } = render(<DashboardLayout />);

    expect(container).toMatchSnapshot();
  });

  it('should open modal on click of add line chart button ', async () => {
    const { getByText } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line chart/i);
    fireEvent.click(addWidgetButton);

    await act(async () => {
      expect(getByText(/Modal Open/i)).toHaveTextContent(/true/i);
    });
  });

  it('should closed modal on click of cancel button ', async () => {
    const { getByText, queryByText } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line Chart/i);
    fireEvent.click(addWidgetButton);

    const cancelButton = getByText(/Cancel/i);
    fireEvent.click(cancelButton);

    await act(async () => {
      expect(queryByText(/Modal Open/i)).toBeNull();
    });
  });

  it('should add new widget with line chart', async () => {
    const { getByText, getByTestId } = render(<DashboardLayout />);

    const addWidgetButton = getByText(/Line Chart/i);
    fireEvent.click(addWidgetButton);

    const okButton = getByText(/Ok/i);
    fireEvent.click(okButton);

    const widget = getByTestId('widget-0');

    await act(async () => {
      expect(widget).toBeInTheDocument();
      expect(widget).toMatchSnapshot();
    });
  });

  it('should save dashboard on click of save dashboard button with empty widgets', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);
    const expectedDashboardData = {
      widgets: [],
      layout: [],
      dashboardId: null,
      name: 'dashboard1',
      count: 0,
    };

    await act(async () => {
      fireEvent.click(saveDashboard);
    });

    expect(api.saveDashboard).toHaveBeenCalledWith(expectedDashboardData);
  });

  it('should save dashboard with widgets on click of save dashboard button ', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);

    const expectedDashboardData = {
      widgets: [
        {
          config: 'config',
          chartType: 'barChart',
          layout: { i: 'widget-0', x: 0, y: Infinity, w: 6, h: 2 },
        },
      ],
      layout: [{ w: 6, h: 2, x: 0, y: 0, i: 'widget-0', moved: false, static: false }],
      dashboardId: null,
      name: 'dashboard1',
      count: 1,
    };

    const addWidgetButton = getByText(/Bar Chart/i);
    fireEvent.click(addWidgetButton);

    const okButton = getByText(/Ok/i);
    fireEvent.click(okButton);

    await act(async () => {
      fireEvent.click(saveDashboard);
    });

    expect(api.saveDashboard).toHaveBeenCalledWith(expectedDashboardData);
  });

  it('should show message for successful saving of dashboard', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);

    await act(async () => {
      fireEvent.click(saveDashboard);
    });

    expect(getByText(/Dashboard dashboard1 Saved Successfully/)).toBeInTheDocument();
  });

  it('should show error while saving dashboard failed', async () => {
    const { getByText } = render(<DashboardLayout />);
    const saveDashboard = getByText(/Save Dashboard/i);
    api.saveDashboard.mockRejectedValue('error');

    await act(async () => {
      fireEvent.click(saveDashboard);
    });

    expect(getByText(/Failed to save dashboard/)).toBeInTheDocument();
  });
});
