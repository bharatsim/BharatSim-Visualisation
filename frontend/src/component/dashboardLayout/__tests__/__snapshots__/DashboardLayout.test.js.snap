// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DashboardLayout /> should add new widget with line chart 1`] = `
<div
  class="react-grid-item react-draggable cssTransforms react-resizable"
  data-grid="[object Object]"
  data-testid="widget-0"
  style="transform: translate(10px,10px); width: -15px; height: 310px; position: absolute;"
>
  <div
    style="height: 100%; width: 100%; box-sizing: border-box; padding: 10px;"
  >
    <div>
      Line Chart
      <span>
        config = config
      </span>
    </div>
  </div>
  <span
    class="react-resizable-handle react-resizable-handle-se"
  />
</div>
`;

exports[`<DashboardLayout /> should match a snapshot for <DashboardLayout /> at initial state 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div>
        File Upload
        <span />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-4"
    >
      <div
        class="MuiBox-root MuiBox-root-6 makeStyles-buttonGroup-5"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Line Chart
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Bar Chart
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-7"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            save dashboard
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="react-grid-layout DashboardLayout-reactGridLayout-1"
      style="height: 10px;"
    />
  </div>
</div>
`;

exports[`<DashboardLayout /> should match a snapshot for <DashboardLayout /> when dashboard is loaded with data 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-9"
  >
    <div
      class="MuiBox-root MuiBox-root-10"
    >
      <div>
        File Upload
        <span />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-11"
    >
      <div
        class="MuiBox-root MuiBox-root-13 makeStyles-buttonGroup-12"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Line Chart
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Bar Chart
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-14"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            save dashboard
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="react-grid-layout DashboardLayout-reactGridLayout-8"
      style="height: 330px;"
    >
      <div
        class="react-grid-item react-draggable cssTransforms react-resizable"
        data-grid="[object Object]"
        data-testid="widget-0"
        style="transform: translate(10px,10px); width: -12px; height: 310px; position: absolute;"
      >
        <div
          style="height: 100%; width: 100%; box-sizing: border-box; padding: 10px;"
        >
          <div>
            Bar Chart
            <span>
              config = config
            </span>
          </div>
        </div>
        <span
          class="react-resizable-handle react-resizable-handle-se"
        />
      </div>
    </div>
  </div>
</div>
`;
